proposed/event-dispatcher.md

```diff
@@ -137,4 +137,132 @@ It is RECOMMENDED that a Notifier and Processor be implemented as distinct objec
+### General event interfaces
+
+```php
+namespace Psr\EventDispatcher;
+
+/**
+ * Marker interface indicating an event instance.
+ *
+ * Event instances may contain zero methods, or as many methods as they
+ * want. The interface MUST be implemented, however, to provide type-safety
+ * to both listeners as well as listener providers.
+ */
+interface EventInterface
+{
+}
+```
+
+```php
+namespace Psr\EventDispatcher;
+
+/**
+ * Mapper from an event to the listeners that are applicable to that event.
+ */
+interface ListenerProviderInterface
+{
+    /**
+     * @param EventInterface $event
+     *   An event for which to return the relevant listeners.
+     * @return iterable[callable]
+     *   An iterable (array, iterator, or generator) of callables.  Each
+     *   callable MUST be type-compatible with $event.
+     */
+    public function getListenersForEvent(EventInterface $event) : iterable;
+}
+```
+
+### Message Interfaces
+
+```php
+namespace Psr\EventDispatcher;
+
+
+/**
+ * This is a marker interface used to identify message events.
+ */
+interface MessageInterface extends EventInterface
+{
+}
+```
+
+```php
+namespace Psr\EventDispatcher;
+
+/**
+ * Defines a notifier for message events.
+ */
+interface MessageNotifierInterface
+{
+    /**
+     * Notify listeners of a message event.
+     *
+     * This method MAY act asynchronously.  Callers SHOULD NOT
+     * assume that any action has been taken when this method
+     * returns.
+     *
+     * @param MessageInterface $event
+     *   The event to notify listeners of.
+     */
+    public function notify(MessageInterface $event) : void;
+}
+```
+
+### Task Interfaces
+
+namespace Psr\EventDispatcher;
+
+```php
+/**
+ * This is a marker interface used to identify task events.
+ */
+interface TaskInterface extends EventInterface
+{
+}
+```
+
+```php
+namespace Psr\EventDispatcher;
+
+/**
+ * A Task whose processing may be interrupted when the task is complete.
+ *
+ * A Processor implementation MUST check to determine if a Task
+ * is marked as stopped after each listener is called.  If it is then it should
+ * return immediately without calling any further Listeners.
+ */
+interface StoppableTaskInterface extends TaskInterface
+{
+    /**
+     * Is propagation stopped?
+     *
+     * This will typically only be used by the Processor to determine if the
+     * previous listener halted propagation.
+     *
+     * @return bool
+     *   True if the Task is complete and no further listeners should be called.
+     *   False to continue calling listeners.
+     */
+    public function isPropagationStopped() : bool;
+}
+```
+
+```php
+/**
+ * Defines a processor for task events.
+ */
+interface TaskProcessorInterface
+{
+    /**
+     * Provide all listeners with a task event to process.
+     *
+     * @param TaskInterface $event
+     *  The task to process.
+     *
+     * @return TaskInterface
+     *  The task that was passed, now modified by callers.
+     */
```

Comment:
Should we write down the exception handling?

Reply:
We don't elsewhere.  I don't want to get into too much detail in the docblocks.

Reply:
K

Reply:
Other PSRs do it BTW:
 * https://github.com/php-fig/fig-standards/blame/master/accepted/PSR-11-container.md#L85-L86
 * https://github.com/php-fig/fig-standards/blame/master/accepted/PSR-16-simple-cache.md#L149
 * https://github.com/php-fig/fig-standards/blame/master/accepted/PSR-17-http-factory.md#L136-L137
 * https://github.com/php-fig/fig-standards/blame/master/accepted/PSR-6-cache.md#L287
 * https://github.com/php-fig/fig-standards/blame/master/accepted/PSR-7-http-message.md#L724
 * etc..

Reply:
If there's a known exception that the object will throw, yes.  But in this case it could literally throw anything, or rather, it's not allowed to prevent anything that gets thrown from getting caught.  `@throws \Throwable Could be anything.` doesn't seem particularly useful.

Reply:
Oh yeah, so then that's the right way to proceed IMHO. I think PSR-15 does the same.

Reply:
That seems... silly, honestly.  Any function that calls any function could conceivably throw anything, or rather not catch it.  Yet nobody (AFAIK) puts `@throws \Throwable` on every function in their API. Only those exceptions that are directly relevant to and generated by the library.

Reply:
Sorry, I meant that PSR-15 did as you are doing, so all fine! :+1: